angr/procedures/posix/dup.py

```diff
@@ -0,0 +1,53 @@
+import angr
+from angr.sim_type import SimTypeInt, SimTypeFd
+from angr.state_plugins import SimSystemPosix
+from angr.storage.file import SimFileDescriptor
+
+##################################
+# dup
+##################################
+
+
+class dup(angr.SimProcedure):  #pylint:disable=W0622
+    def run(self, oldfd): # pylint:disable=arguments-differ
+        if oldfd not in self.state.posix.fd.keys():
+            return self.state.libc.ret_errno('EBADF')
+
+        # The new fd gets the lowest free number, so we search
+        newfd = len(self.state.posix.fd.keys()) # e.g. '3' for [0, 1, 2]
+        for i, fd in enumerate(self.state.posix.fd.keys()):
```

Comment:
this needs to be enumerate(sorted(keys)) in order for this algorithm to work. tbh I would just scan upwards from 0 and pick the first free one.

Reply:
Done.